### PR TITLE
esdt improvements refactor

### PR DESCRIPTION
**Pull request by BeniaminDrasovean**
This refactor has two pourposes:

1. If a token type is migrated (the token type is correctly set inside the esdtData), do not access the `globalSettingsHandler`, which saves it's data inside the system account
2. Also save the token type in the esdtData for SFTs & metaESDTs

### CODE REVIEW

File: builtInFunctions/esdtDataStorage.go [hunk 1]
```diff
@@ -411,8 +414,21 @@ func (e *esdtDataStorage) SaveESDTNFTToken(
...
+		// it means that the token type was already migrated.
+		return true
+	}
+
+	return false
+}
+
 // removeNFTMetadataFromSystemAccountIfNeeded removes the NFT metadata from the system account if needed
-func (e *esdtDataStorage) removeNFTMetadataFromSystemAccountIfNeeded(esdtTokenKey []byte, nonce uint64, esdtData *esdt.ESDigitalToken) error {
+func (e *esdtDataStorage) migrateTokenTypeAndMetadataIfNeeded(esdtTokenKey []byte, nonce uint64, esdtData *esdt.ESDigitalToken) error {
```
Comment (ssd04) [inline builtInFunctions/esdtDataStorage.go:431]: update or remove comment

File: builtInFunctions/esdtDataStorage.go [hunk 2]
```diff
@@ -421,14 +437,13 @@ func (e *esdtDataStorage) removeNFTMetadataFromSystemAccountIfNeeded(esdtTokenKe
 	if err != nil {
 		return err
 	}
-	if esdtData.Type != uint32(core.NonFungible) || tokenType != uint32(core.NonFungibleV2) {
+	if tokenType == esdtData.Type {
 		return nil
 	}
+	esdtData.Type = tokenType
```
Comment (ssd04) [inline builtInFunctions/esdtDataStorage.go:443]: can we have a unit test for sft, metaesdt to check that the correct type is set?

File: builtInFunctions/esdtDataStorage_test.go [hunk 3]
```diff
@@ -615,19 +615,12 @@ func TestEsdtDataStorage_SaveESDTNFTToken(t *testing.T) {
...
 		key := baseESDTKeyPrefix + tokenIdentifier
 		tokenKey := append([]byte(key), big.NewInt(int64(nonce)).Bytes()...)
-		setTokenTypeCalled := false
 
 		args := createMockArgsForNewESDTDataStorage()
 		args.GlobalSettingsHandler = &mock.GlobalSettingsHandlerStub{
 			GetTokenTypeCalled: func(esdtTokenKey []byte) (uint32, error) {
 				return uint32(core.NonFungibleV2), nil
 			},
-			SetTokenTypeCalled: func(esdtTokenKey []byte, tokenType uint32) error {
```
Comment (sasurobert) [inline builtInFunctions/esdtDataStorage_test.go:625]: why did you delete the test ?

File: builtInFunctions/esdtDataStorage.go (hunk 1 above)
Comment (sasurobert) [reply builtInFunctions/esdtDataStorage.go:431]: +1

Comment (BeniaminDrasovean) [reply builtInFunctions/esdtDataStorage.go:431]: Updated the comment. It now states the reason why this function is needed.

File: builtInFunctions/esdtDataStorage.go (hunk 2 above)
Comment (BeniaminDrasovean) [reply builtInFunctions/esdtDataStorage.go:443]: Added.

File: builtInFunctions/esdtDataStorage_test.go (hunk 3 above)
Comment (BeniaminDrasovean) [reply builtInFunctions/esdtDataStorage_test.go:625]: This call was redundant. In the normal flow, `SetTokenType` is called by `esdtSetTokenType` and the token type is set inside `globalSettingsHandler`. Here we should set the type from `globalSettingsHandler` inside `esdtData` in order to be consistent, so no `SetTokenTypeCalled` is needed

**Review by sasurobert — DISMISSED** [review]

**Review by ssd04 — APPROVED** [review]

File: builtInFunctions/esdtDataStorage.go [hunk 4]
```diff
@@ -411,8 +414,21 @@ func (e *esdtDataStorage) SaveESDTNFTToken(
...
 }
 
-// removeNFTMetadataFromSystemAccountIfNeeded removes the NFT metadata from the system account if needed
-func (e *esdtDataStorage) removeNFTMetadataFromSystemAccountIfNeeded(esdtTokenKey []byte, nonce uint64, esdtData *esdt.ESDigitalToken) error {
+func (e *esdtDataStorage) checkTheCorrectTokenTypeIsSet(esdtData *esdt.ESDigitalToken) bool {
+	if !e.enableEpochsHandler.IsFlagEnabled(DynamicEsdtFlag) {
+		return true
+	}
+
+	if esdtData.Type != uint32(core.NonFungible) {
```
Comment (ssd04) [inline builtInFunctions/esdtDataStorage.go:422]: shouldn't we put these 2 checks together? if dynamic esdt flag is enabled it doesn't really mean that the token type was updated

File: builtInFunctions/esdtGlobalSettings_test.go [hunk 5]
```diff
@@ -296,3 +298,116 @@ func TestESDTGlobalSettingsBurnForAll_ProcessBuiltInFunction(t *testing.T) {
...
+			core.BuiltInFunctionESDTPause,
+			falseHandler,
+		)
+
+		acnt.Storage["key"] = []byte{byte(notSet)}
+		val, err := globalSettingsFunc.GetTokenType([]byte("key"))
+		require.Nil(t, err)
+		require.Equal(t, uint32(core.NonFungible), val)
+	})
+	t.Run("retrieve token type error ", func(t *testing.T) {
```
Comment (ssd04) [inline builtInFunctions/esdtGlobalSettings_test.go:371]:
```suggestion
	t.Run("retrieve token type error", func(t *testing.T) {
```

File: builtInFunctions/esdtGlobalSettings_test.go [hunk 6]
```diff
@@ -296,3 +298,116 @@ func TestESDTGlobalSettingsBurnForAll_ProcessBuiltInFunction(t *testing.T) {
...
+			core.BuiltInFunctionESDTPause,
+			falseHandler,
+		)
+
+		acnt.Storage["key"] = []byte{0, 100}
+		val, err := globalSettingsFunc.GetTokenType([]byte("key"))
+		require.True(t, strings.Contains(err.Error(), "invalid esdt type"))
+		require.Equal(t, uint32(0), val)
+	})
+	t.Run("convert to esdt token type error", func(t *testing.T) {
```
Comment (ssd04) [inline builtInFunctions/esdtGlobalSettings_test.go:392]:
```suggestion
	t.Run("convert to esdt token type", func(t *testing.T) {
```
?

File: builtInFunctions/esdtDataStorage.go (hunk 4 above)
Comment (BeniaminDrasovean) [reply builtInFunctions/esdtDataStorage.go:422]: If the tokent type is `!= core.NonFungible` it means that for sure it was updated. I think it is more obvious and easy to understand as 2 separate checks.

**Review by ssd04 — APPROVED** [review]

**Review by sasurobert — APPROVED** [review]
